### PR TITLE
THU-285: [PART 3] E2E encryption — backend API endpoints & device logic update

### PR DESCRIPTION
**Pull request by raivieiraadriano92**
## Summary

Implements all encryption API endpoints and updates existing device logic to use the `status` field.

### New endpoints

**`POST /devices`** — Register or identify a device
- Input: `{ deviceId, publicKey, name? }`
- If device exists + TRUSTED → returns `{ status: "TRUSTED", envelope }`
- If device exists + APPROVAL_PENDING → returns `{ status: "APPROVAL_PENDING", firstDevice }`
- If new device → creates with APPROVAL_PENDING, returns `{ status: "APPROVAL_PENDING", firstDevice }`
- `firstDevice: true` when no envelopes exist for the user (FE generates CK locally)

**`POST /devices/:deviceId/envelope`** — Store envelope + mark trusted
- Input: `{ wrappedCK, canaryIv?, canaryCtext? }`
- Stores envelope, optionally stores canary (idempotent), marks device TRUSTED
- Callable by the device itself (first device) or by a trusted device (approval)

**`GET /devices/me/envelope`** — Fetch own envelope
- Requires `X-Device-ID` header
- Returns calling device's envelope only (enforced server-side)

**`GET /encryption/canary`** — Fetch canary for recovery key verification
- Returns `{ canaryIv, canaryCtext }` for the current user

### Updated existing logic

**`POST /account/devices/:id/revoke`** — now also:
- Deletes the device's envelope from `envelopes` table
- Sets `status = 'REVOKED'` (in addition to existing `revokedAt`)

**PowerSync token validation** — now checks `status === 'REVOKED'` in addition to `revokedAt`

### DAL changes

- `getDeviceById` now returns `status` field
- `revokeDevice` now sets both `status: 'REVOKED'` and `revokedAt`
- New: `registerDevice` — creates device with APPROVAL_PENDING + public key
- New: `updateDeviceStatus` — updates device status
- New encryption DAL: `getEnvelopeByDeviceId`, `hasEnvelopesForUser`, `upsertEnvelope`, `deleteEnvelope`, `getEncryptionMetadata`, `insertEncryptionMetadataIfNotExists`

### Files changed
- `backend/src/api/encryption.ts` — new route file (4 endpoints)
- `backend/src/api/account.ts` — revoke deletes envelope
- `backend/src/api/powersync.ts` — status check
- `backend/src/dal/encryption.ts` — new DAL
- `backend/src/dal/devices.ts` — updated + new functions
- `backend/src/dal/index.ts` — re-exports
- `backend/src/index.ts` — registers encryption routes

## Test plan

- [ ] `POST /devices` with new deviceId + no envelopes → `{ status: "APPROVAL_PENDING", firstDevice: true }`
- [ ] `POST /devices` with new deviceId + envelopes exist → `{ status: "APPROVAL_PENDING", firstDevice: false }`
- [ ] `POST /devices` with existing TRUSTED device → `{ status: "TRUSTED", envelope: "..." }`
- [ ] `POST /devices` with existing PENDING device → `{ status: "APPROVAL_PENDING", firstDevice }`
- [ ] `POST /devices/:id/envelope` with wrappedCK + canary → stores both, returns `{ status: "TRUSTED" }`
- [ ] `POST /devices/:id/envelope` without canary (approval path) → stores envelope only
- [ ] `GET /devices/me/envelope` with valid X-Device-ID → returns envelope
- [ ] `GET /devices/me/envelope` without header → 400
- [ ] `GET /encryption/canary` → returns canary for authenticated user
- [ ] `GET /encryption/canary` with no encryption setup → 404
- [ ] `POST /account/devices/:id/revoke` → deletes envelope + sets status REVOKED
- [ ] PowerSync token request from REVOKED device → 403
- [ ] All endpoints return 401 without session
- [ ] `bun run typecheck` passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Introduces new authenticated endpoints and transactional writes around device trust/envelope storage, so bugs could impact device authorization, key material persistence, or revocation enforcement. Changes are scoped to device/encryption flows but touch PowerSync token validation and account revoke behavior.
> 
> **Overview**
> Adds a new `Encryption API` surface (`POST /devices`, `POST /devices/:deviceId/envelope`, `GET /devices/me/envelope`, `GET /encryption/canary`) to support E2E setup: device registration, envelope storage/approval with transactional checks, and canary retrieval.
> 
> Updates device handling to rely on an explicit `status` field: `getDeviceById` now returns status, `upsertDevice` inserts as `APPROVAL_PENDING`, `revokeDevice` sets `status='REVOKED'` (and `revokedAt`), and PowerSync token issuance now blocks devices with `status='REVOKED'`.
> 
> Tightens revocation semantics by making `POST /account/devices/:id/revoke` delete the device’s envelope and revoke in a single DB transaction, and adds/extends backend tests (including a PGlite rerun-safe ID strategy) to cover encryption flows and revoke edge cases.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 7ba9c0f6f3b54f16422b82f964a5ef11862ff9ee. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: backend/src/api/encryption.ts [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        const device = await getDeviceById(database, deviceId)
+        if (!device || device.userId !== userId) {
+          set.status = 404
+          return { error: 'Device not found' }
+        }
+
+        if (device.status === 'REVOKED') {
+          set.status = 403
+          return { error: 'Device has been revoked' }
+        }
```
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/encryption.ts:116]: ⚡ The caller-trust check was already added in a prior commit (lines 118-137 verify X-Device-ID header and check caller is TRUSTED). This latest push further hardens the endpoint by wrapping the entire flow in a transaction for atomicity.

File: backend/src/api/account.ts [hunk 2]
```diff
@@ -27,6 +27,7 @@ export const createAccountRoutes = (auth: Auth, database: typeof DbType) => {
     })
     .post('/devices/:id/revoke', async ({ params, set, user: sessionUser }) => {
       const userId = sessionUser!.id
+      await deleteEnvelope(database, params.id, userId)
       await revokeDevice(database, params.id, userId)
```
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/account.ts]: ⚡ Good catch — this was already wrapped in a database transaction in a prior commit (lines 30-34 of account.ts use database.transaction).

File: backend/src/api/encryption.ts [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        const { wrappedCK, canaryIv, canaryCtext } = body
+
+        // Verify the target device exists and belongs to this user
+        const device = await getDeviceById(database, deviceId)
+        if (!device || device.userId !== userId) {
+          set.status = 404
+          return { error: 'Device not found' }
+        }
+
+        if (device.status === 'REVOKED') {
```
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/encryption.ts]: ⚡ Good call — done in the latest push. Added a 409 Conflict guard that rejects envelope writes to already-TRUSTED devices unless the caller is the device itself (self re-key).

File: backend/src/api/encryption.ts [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+        // Verify caller is a TRUSTED device (or this is a first-device bootstrap)
+        const callerDeviceId = request.headers.get('x-device-id')?.trim()
+        if (!callerDeviceId) {
+          set.status = 400
+          return { error: 'X-Device-ID header is required' }
+        }
+
+        const envelopesExist = await hasEnvelopesForUser(database, userId)
+        const isFirstDeviceBootstrap = !envelopesExist && callerDeviceId === deviceId
```
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/encryption.ts]: ⚡ Good call — done in the latest push. The hasEnvelopesForUser check and all subsequent writes (upsertEnvelope, insertEncryptionMetadataIfNotExists, updateDeviceStatus) are now wrapped in a single database transaction, preventing the concurrent first-device bootstrap race condition.

File: backend/src/api/account.ts [hunk 5]
```diff
@@ -27,6 +27,7 @@ export const createAccountRoutes = (auth: Auth, database: typeof DbType) => {
     })
     .post('/devices/:id/revoke', async ({ params, set, user: sessionUser }) => {
       const userId = sessionUser!.id
+      await deleteEnvelope(database, params.id)
```
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/account.ts]: Fixed — `deleteEnvelope` in `dal/encryption.ts` uses both `deviceId` and `userId` in the WHERE clause. The revoke endpoint is also now wrapped in a transaction (lines 30-34 of account.ts).

File: backend/src/api/encryption.ts [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        if (canaryIv && canaryCtext) {
+          await insertEncryptionMetadataIfNotExists(database, {
+            userId,
+            canaryIv,
+            canaryCtext,
+          })
+        }
+
+        // Mark device as trusted
+        await updateDeviceStatus(database, deviceId, userId, 'TRUSTED')
```
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/encryption.ts]: Fixed — the endpoint checks `device.status === 'REVOKED'` and returns 403 (line 113).

File: backend/src/api/encryption.ts [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+          userId,
+          name: deviceName,
+          publicKey,
+        })
+
+        const envelopesExist = await hasEnvelopesForUser(database, userId)
+        return {
+          status: 'APPROVAL_PENDING' as const,
+          firstDevice: !envelopesExist,
+        }
```
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/encryption.ts:89]: Fixed — `POST /devices` now explicitly handles REVOKED status with a 403 response (lines 70-73).

File: backend/src/api/powersync.ts [hunk 8]
```diff
@@ -37,7 +37,7 @@ const validateDeviceNotRevoked = async (
     if (deviceRow.userId !== userId) {
       return { ok: false, status: 409, body: { code: 'DEVICE_ID_TAKEN' } }
     }
-    if (deviceRow.revokedAt != null) {
+    if (deviceRow.status === 'REVOKED' || deviceRow.revokedAt != null) {
```
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/powersync.ts:40]: Fixed in PR #496 (db63698e) — added `status` and `public_key` to `uploadDenyColumns` for the devices table.

File: backend/src/api/encryption.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/encryption.ts:116]: Fixed — the envelope endpoint now verifies the X-Device-ID header and checks the caller is TRUSTED (lines 118-147), with a first-device bootstrap exception. All wrapped in a transaction.

File: backend/src/api/account.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/account.ts]: Fixed — revoke endpoint is now wrapped in `database.transaction()` (lines 30-34 of account.ts).

File: backend/src/api/encryption.ts [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
```
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/encryption.ts:89]: The register endpoint now handles all device states explicitly (TRUSTED returns envelope, APPROVAL_PENDING returns status, REVOKED returns 403, conflict returns 409). New devices are registered with APPROVAL_PENDING status.

File: backend/src/api/encryption.ts [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+          }
+
+          // Already pending — return status + firstDevice hint
+          if (existingDevice.status === 'APPROVAL_PENDING') {
+            const envelopesExist = await hasEnvelopesForUser(database, userId)
+            return {
+              status: 'APPROVAL_PENDING' as const,
+              firstDevice: !envelopesExist,
+            }
+          }
```
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/encryption.ts:67]: By design — re-registration for an existing device returns its current status without modifying the device. The public key is only set during initial registration. Existing TRUSTED devices return their envelope, APPROVAL_PENDING devices return their status.

File: backend/src/api/encryption.ts [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+          // Revoked — device cannot re-register
+          if (existingDevice.status === 'REVOKED') {
+            set.status = 403
+            return { error: 'Device has been revoked' }
+          }
```
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/encryption.ts:73]: Fixed in PR #496 (db63698e) — `status` and `public_key` added to `uploadDenyColumns`.

File: backend/src/api/encryption.ts [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+            return { error: 'Device ID already taken' }
+          }
+
+          // Already trusted — return envelope
+          if (existingDevice.status === 'TRUSTED') {
+            const envelope = await getEnvelopeByDeviceId(database, deviceId, userId)
+            return {
+              status: 'TRUSTED' as const,
+              envelope: envelope?.wrappedCk ?? null,
+            }
```
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/encryption.ts:57]: By design — the PowerSync `upsertDevice` creates devices with default TRUSTED status for backward compatibility with the existing non-encrypted sync flow. The encryption registration endpoint (`POST /devices`) creates with APPROVAL_PENDING. These are separate code paths.

File: backend/src/api/encryption.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/encryption.ts]: Fixed — added a 409 Conflict guard (lines 125-129) that rejects envelope writes to already-TRUSTED devices unless the caller is the device itself (self re-key).

File: backend/src/api/encryption.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/encryption.ts]: Fixed — the `hasEnvelopesForUser` check and all subsequent writes are now wrapped in a single `database.transaction()` (lines 131-167), preventing concurrent first-device bootstrap.

File: backend/src/api/encryption.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/api/encryption.ts:57]: Fixed in PR #496 (e816bfe9) — changed the schema default from `TRUSTED` to `APPROVAL_PENDING` so no code path can accidentally create a trusted device without going through the approval flow. Existing devices are backfilled to `TRUSTED` via migration.

File: backend/src/dal/devices.ts [hunk 13]
```diff
@@ -2,33 +2,75 @@ import type { db as DbType } from '@/db/client'
...
-/** Upsert a device: insert new or update lastSeen/name for existing. Only updates if userId matches. */
+/** Upsert a device: insert new with APPROVAL_PENDING or update lastSeen/name for existing. Only updates if userId matches. */
 export const upsertDevice = async (
   database: typeof DbType,
   device: { id: string; userId: string; name: string; lastSeen: Date; createdAt: Date },
 ) =>
   database
     .insert(devicesTable)
-    .values(device)
+    .values({ ...device, status: 'APPROVAL_PENDING' })
```
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/dal/devices.ts:25]: fixed in [e816bfe](https://github.com/thunderbird/thunderbolt/pull/496/commits/e816bfe90b722f634a40be5449f430e924a42e28)

File: backend/src/dal/devices.ts [hunk 14]
```diff
@@ -2,33 +2,75 @@ import type { db as DbType } from '@/db/client'
...
+      name: device.name,
+      publicKey: device.publicKey,
+      status: 'APPROVAL_PENDING',
+      createdAt: new Date(),
+      lastSeen: new Date(),
+    })
+    .onConflictDoUpdate({
+      target: devicesTable.id,
+      set: { publicKey: device.publicKey, lastSeen: new Date() },
+      setWhere: eq(devicesTable.userId, device.userId),
```
[the remark this answers is not in the document]
Comment (raivieiraadriano92) [reply backend/src/dal/devices.ts:74]: The only theoretical scenario is the same user sending two requests with the same device ID but different public keys simultaneously — which would require regenerating the key pair between requests, not a realistic client behavior.
